### ISSUE TITLE
Trailing whitespace in `MATH` and `KernClass2` lines

### ISSUE THREAD

**Issue by waldyrious** [fontforge/fontforge#4265]
Previous instances of trailing whitespace produced by FontForge have been fixed in #1658 and #2713 (/cc @adrientetar and @jre-wine), but it seems that this still happens in some `MATH` and `KernClass2` lines, as shown in [this diff](https://github.com/alif-type/libertinus/pull/316#issuecomment-609391396).

I would expect the fixes to involve editing around [this line](https://github.com/fontforge/fontforge/blob/master@%7B2020-04-04%7D/fontforge/sfd.c#L2784) and [this one](https://github.com/fontforge/fontforge/blob/master@%7B2020-04-04%7D/fontforge/sfd.c#L2540), but I'm not sufficiently familiar with the codebase to know what exactly the fix would be, so I thought I'd open an issue instead.

### Checklist
- [x] Screenshot (I suppose the diff fills this role)
- [x] The FontForge version and the operating system you're using:
      macOS catalina; FontForge 20190801 (build date: 12:50 UTC 14-Nov-2019)
- [x] The behavior you expect to see, and the actual behavior
- [ ] Steps to reproduce the behavior
- [ ] \(optional) Possible solution/fix/workaround

**Comment by ctrlcctrlv** [comment]
If FontForge understands the file, it's valid. SFD is not made to be consumable by other programs. Trailing whitespace is valid, like in most other formats.

You should use an [SFD normalizer](https://github.com/alif-type/libertinus/blob/master/tools/sfdnormalize.py) for Git purposes, similar to what the Libertinus project does.

Just because cosmetic issues like this were patched 4 to 6 years ago doesn't mean they will be now. sfdnormalize.py is now very good, spending developer time on making SFD files more revision control-friendly is needlessly duplicated effort.

Obviously a PR would be accepted but bugs like this should not be opened unless you plan to open a PR too.

**Comment by waldyrious** [comment]
I would be happy to submit a PR if I'm provided some guidance regarding how to implement this. I mean, I suppose that I can't simply remove the `putc(' ',sfd);` lines. Any pointers would be most welcome.

**Comment by ctrlcctrlv** [comment]
Read the source...

For MATH, SFDDumpDeviceTable writes out nothing if `devtab` is NULL, leading to the trailing whitespace. You could just check that before calling it, as line 1513 does.

To be quite honest the code right now is better and simpler than adding a double check for NULL just to avoid some whitespace which is not even out-of-spec. Indeed you've chosen two invocations of SFDDumpDeviceTable to target, but the function is used in many more places. So will all those check for NULL before calling it also? If so, the NULL check should be dropped from SFDDumpDeviceTable.

Writing a space cannot just be added to SFDDumpDeviceTable, as some invocations come after a space and some don't. _Unless_ you prove via adequate testing that the places where spaces aren't written before the device table can have spaces and old FontForge versions will understand that.

The PRs you cited dealt mostly with SFDDumpUTF7Str and changing format strings so were more acceptable.

I reiterate that you need to incorporate a normalizer into your workflow and that this is a dubious place to put anyone's time.

**Comment by waldyrious** [comment]
Thanks so much for the detailed explanation. I did suspect that such complexities might arise if this were to be fixed. The good news is that, precisely in the context that called my attention to this issue, I have just recently contributed a whitespace trimming feature to Libertinus' normalizer (https://github.com/alif-type/libertinus/pull/316/commits/2d0ded122989b52a5d7b3395663f30659b201d5f), so hopefully others can now use that tool to work around this (admittedly minor) issue.

I'll go ahead and close this issue, as I agree with your suggestion that this fix would consume disproportionately more time and effort than the final effect would warrant.
